src/utils/partition_stl_containers.hpp

```diff
@@ -0,0 +1,83 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//=======================================================================================
+#ifndef UTILS_PARTITION_STL_CONTAINERS_HPP_
+#define UTILS_PARTITION_STL_CONTAINERS_HPP_
+
+#include <string>
+#include <vector>
+
+#include "error_checking.hpp"
+
+namespace parthenon {
+namespace Partition {
+// TODO(JMM): The templated type safety with T* and T may need to be
+// changed if we move to sufficiently general container objects.
+template <typename T>
+using Partition_t = std::vector<std::vector<T *>>;
+
+// x/y rounded up
+// See discussion here for limitations and alternatives
+// https://stackoverflow.com/questions/2745074/fast-ceiling-of-an-integer-division-in-c-c
+KOKKOS_INLINE_FUNCTION
+int IntCeil(int x, int y) {
+  PARTHENON_DEBUG_REQUIRE(x > 0 && y > 0, "ceil only works for x,y > 0");
+  // avoids overflow in x+y
+  return 1 + ((x - 1) / y);
+}
+
+// Takes container of elements and fills partitions
+// of size N with pointers to elements.
+// Assumes Container_t has STL-style iterators defined
+template <typename Container_t, typename T>
+void ToSizeN(Container_t &container, const int N, Partition_t<T> &partitions) {
+  using std::to_string;
+
+  PARTHENON_REQUIRE_THROWS(N > 0, "You must have at least 1 partition");
+  int nelements = container.size();
+  PARTHENON_REQUIRE_THROWS(nelements > 0,
+                           "You must have at least 1 element to partition");
+  std::string msg = ("Cannot partition " + to_string(nelements) +
+                     " elements into partitions of size " + to_string(N) + ".");
+  PARTHENON_REQUIRE_THROWS(nelements >= N, msg);
```

Comment:
Hm... I don't think I like any of these checks. I think the algorithm otherwise produces perfectly reasonable results for size-0 containers and for containers that are smaller than the partition size.

I think we already have a check that each rank gets at least 1 block, and that's going to give the user a more useful error message. So don't complicate the algorithm and tests for this - just allow passing a container of size 0.

I don't think we should require the first partition to be filled - if it's determined that a specific package benefits from a precise partition size, it shouldn't fail out just because in some instance a partition didn't get filled up. I think what you end up if you do that is everybody guarding their calls to `ToSizeN` with `N=std::min(container.size(), N)`. Consider these two scenarios for a chosen partition size of 32: 1) you have a container size of 31, and so it fails, and 2) you have a container size of 33, which gives you two partitions of size 32 and 1. Arguably 2) is worse because you probably would have been better off with a single partition of 33, but we allow 2) and disallow 1).

Reply:
(Oh, sorry, the N > 0 check is just fine, it's the rest of them that I'd remove)

Reply:
Ah good point. I'll remove those checks.